source/develop/release-management/features/virt/vm-lifecycle-with-kubevirt.md

```diff
@@ -0,0 +1,130 @@
+4. Most of the VMs monitoring remains the same.  
+5. Most of the VM related operations are still done directly with VDSM (rather than through Kubervirt).
+6. If we fail to run a VM on a node several times in a short-period of time, the node should be non-operational (and thus not be scheduled until fixed).
+
+## Milestone 3: Dynamic Cluster State Managed by Kubevirt
+
+### Goal
+At this point we have basic monitoring of VMs. This milestone is about detecting more changes in running VMs in Kubevirt and report them to oVirt. Ideally, most of VmDynamic would be managed in Kubevirt and will be in oVirt only for display purpose. This allows us to switch VM management operations (shutdown, migrate, suspend, ..) to be done via Kubervirt.
+
+### Design
+1. Receive the VMs state from the node and updating the dynamic part in the VM representation.  
+2. Starting highly-available VMs in Kubevirt.
+3. Change VM management operations: hibernate, restore, resume, migrate, shutdown, reboot, cancel-migration, power-off.
+
+### Benefits
+1. Kubevirt becomes the owner of the runtime state of the cluster while oVirt manages its static state (host-deployments, VM templates, import VMs from external providers, and so on).  
+2. Many VM operations are delegated to Kubevirt.
+
+## Future Changes
+
+Future changes depend on having the knowledge of storage and network resources in Kubevirt and thus harder to predict.  
+
+## VM Representation
+
+I believe that the best way to represent the VM is in JSON as follow:  
+1. Scheduling data: the cluster-related data mentioned before, in the form of properties in JSON.  
+2. VM dynamic data: things like status of the VM, run_on_vds, console ip, and so on. Initially they should be set with default values. Once the monitoring in Kubevirt receives them from the node, it should update the VM representation and some component needs to listen to these changes and report them to the engine in oVirt. Note that we may want to report some of the dynamic data that exist today, like the applications list, in a different way.   
+3. VM configuration: the static configuration of the VM, including its devices.  
+
+In oVirt we are in the process of representing the VM configuration in XML (domxml) to eliminate the duplicate conversion of the data (once in the engine, converting the internal representation into a dictionary, and once in VDSM, converting the dictionary into domxml). We would prefer to keep this approach and minimize data conversions. In light of this approach, there are few options of how to represent the VM configuration:  
+1. To generate the VM representation in JSON. There is a proposal (on-going work?) to implement another binding for libvirt that will enable to pass the VM definition in JSON. With this, we would still have only one conversion of the VM configuration. That seems like the best approach since Kubernetes works with Yaml formats.  
```

Comment:
Sure, I think a good example is starting a VM **V** with a disk **D** from storage **S**

In oVirt it works basically the following way ( I guess it can differ a little bit depending on the storage used):

oVirt:

 1) Add storage **S** to the engine
 2) engine tells vdsm(s) to attach the storage **S**
 3) engine tells vdsm or storage to create an empty disk
 4) engine generates a domain xml (at least soon it will), and passes it to vdsm
 5) vdsm fills in some missing details into the domain xml, only it can know
 6) vdsm tells libvirt to start the VM

Kubevirt:

 1) Define a storage **S** on Kubevirt (Kubevirt does nothing with them so far)
 2) Define a volume claim **C** for **S** to Kubevirt (again Kubevirt does nothing with it so far)
 3) Add a VM Spec **V** which references the volume claim **C**
 4) Kubevirt schedules the VM **V** (find out the target node and start a container there where the VM should be at the end)
 5) virt-handler on the node now sees that VM **V** needs to be started on the node it is running on (it is not told to start it, it is watching the VMs)
 6) virt-handler sees that the VM **V** needs a volume from claim **C** from storage **S** (which are all not part of the VM spec, it will fetch and examine these resources)
 7) virt-handler mounts (or provides it for passthrough) that storage to the host if not present (so it is lazy, on a need basis)
 8) virt-handler makes sure that the volume to fulfill the volume claim **C** is created if not already present
 9) virt-handler generates the resulting domain xml with all the informations it now has
 10) virt-handler starts the VM

Note that the steps 6) to 10) might not necessarily be done by virt-handler. Some of the work might already be done between 4) and 5) by native kubernetes constructs, but in principle this is how it works.

Also note that the order of defining the dependent components 

For networks it works in a similar fashion.

So really the node is responsible for figuring out everything. For instance if you define a VM which references a non existent claim, the node will tell you that after it tried using it.